### ISSUE TITLE
grid_sample with relative grid

### ISSUE THREAD

**Issue by madshi** [pytorch/pytorch#94816]
### 🚀 The feature, motivation and pitch

Currently, "grid_sample" requires a "grid" with absolute values. That seems fine in theory. However, practically, if we're talking about big images, e.g. 4K video, the absolute grid values will become very large, up to value 3840 on the right most pixel column. Which becomes a big problem if we're talking about a Float16 absolute "grid".

E.g. please play around with Float16 numbers here:

https://evanw.github.io/float-toy/

If you enter 3799, actually it seems that Float16 cannot properly express that number, so we get 3798 (unless that calculator on that website is buggy?). Which means when using a Float16 grid, grid_sample becomes totally unusable because at high image resolutions Float16 becomes so inaccurate that it's basically utterly broken.

Now imagine you want a pixel to move by 0.1 pixels. So the grid for pixel 3800 would say "3800.1". However, Float16 doesn't even have remotely enough precision for that! Hello, artifacts galore.

Now we could argue that we should simply create a Float32 grid. It works well. So why not do that? Well, it consumes more RAM, and it's slower. So it would be much nicer if we could use a Float16 grid.

**So here comes my wish**: There should be a grid_sample alternative which accepts a relative grid. With the "grid" all zero, basically input and output images should be identical. If you want to move a pixel 0.1 to the right, that pixel in the grid would simply have a value of "0.1". Such a relative grid should work perfectly fine with Float16.

Plus, most neural networks which create a "grid" actually calculate relative values natively, anyway! They don't usually calculate absolute values. So if you look through all the public neural networks which use grid_sample, most of them painfully convert their internal relative grid to an absolute grid, to make grid_sample happy. Which is a completely useless conversion, which only makes the use of grid_sample more complex, plus it introduces the Float16 accuracy problem explained above.

Practically, in all my neural networks which use grid_sample, if there was a relative addressing version of grid_sample available, my code would be simpler, and I could stick to Float16, which my neural network natively outputs, anyway. So it would also perform better, with no relative -> absolute conversion needed, and no Float16 -> Float32 conversion needed, either.

**Comment by madshi** [comment]
P.S: I've added a numerical example to the initial feature request text, to explain why a Float16 "grid" parameter is basically unusable (way too inaccurate) with the current grid_sample "absolute" addressing approach.

**Comment by madshi** [comment]
P.P.S: Sorry for the posting flood.

While we're at it, it would be nice if the "relative" grid_sample version would also accept the grid in NCHW order. Currently, for the current grid_sample version, we always have to transpose the "grid" to NHWC order to make grid_sample happy, while the convolutional neural network (which produces the "grid" data) outputs NCHW. It's really weird that grid_sample wants the "input" to be in NCHW order, but the grid in NHWC order! I mean, the transpose is not that big a deal, but it's yet one more unnecessary data shuffling step which eats performance and makes things more complicated than necessary.

**Comment by mikaylagawarecki** [comment]
Hi @madshi, thanks for the suggestion! We would accept a PR that adds the functionality you suggested to `grid_sample` gated behind a new kwarg :)

**Comment by madshi** [comment]
Glad to hear you seem to like the suggestion. Unfortunately, I usually lose 20% of my hair whenever I try to compile open source projects. Being on Windows, recompiling open source projects isn't always easy. For my needs, I've found a good workaround, using "absolute" float32 for training, and "relative" float16 for inference (with a TensorRT plugin and some graphsurgeon magic). So I'm good now, and don't really strictly need this suggestion to be implemented.

Maybe a simpler to implement alternative approach would be to let PyTorch simply output a warning to the terminal whenever a float16 grid is used, so the user is aware that it is likely to cause image quality issues? Should cost you only a couple of minutes to implement, and would at least make the users aware that float32 is needed for the grid.
